Introduce the new versioning options for the SCM/CI integration (#417)
openSUSE/obs-landing: _posts/development/2023-06-23-new-versioning-options-for-the-scm-ci-integration.md

---
layout: post

> title: Enabling Larger Changes in SCM/CI Workflows

Reviewer: ```suggestion
title: SCM/CI Workflow Versions. Making Larger Changes Less Painful For You.
```

category: development
---

> Hello everyone!

Reviewer: nitpick: To me this greeting feels weird in a blog post...

Author: Okay, I'll remove it

Today we're going to explain the versioning system we're using to release new features to the SCM/CI integration.

# Versioning the Workflow Configuration

> We are introducing the versioning as a way to release new features in the SCM/CI without breaking existing user's configuration. Those features are going to be released under a MAJOR.MINOR version scheme. You know the drill already: current version is 1.0, so the next minor version will be 1.1 and the next major version will probably be 2.0.

Author: So better not giving that much detail?

> We are introducing the versioning as a way to release new features in the SCM/CI without breaking existing user's configuration. Those features are going to be released under a MAJOR.MINOR version scheme. You know the drill already: current version is 1.0, so the next minor version will be 1.1 and the next major version will probably be 2.0.

Author: Ah, the `probably`... okay, got it

> A major version update will introduce big features and may bring breaking changes, so it will require adapting your workflow configuration.

Reviewer: ```suggestion
A major version update will introduce big features and may bring breaking changes, so it will require adapting your workflow configuration once you update to a new major version.
```

A minor version update will only include small, backwards-compatible changes. If you don't touch anything, your workflow will still work as before. Only when you make some minor adjustments to your workflow configuration you'll benefit of a new feature.

The workflow version can be set as the top-level element of the workflow configuration yaml file. Like this:

    version: 1.0
    workflow:
        step:
            - link_package:
                source_project: GNOME:Factory
                source_package: gnome-shell
                target_project: home:jane:playground

> We won't enforce any specific version in the yaml configuration. So if you don't set any version, your Workflow configuration will default to the latest minor version (1.1 as of the release of this blog post). That way you can test those minor features at your own pace.

Reviewer: Hm and what version will it be if there are two minor versions? 1.1 and 2.1?

Author: 1.1, because the we would be breaking everybody's config by switching to a major version

Reviewer: I guess you are tying to say: `least` or `lowest` minor version and not `latest`.

Author: I've been thinking about `least`, and I'm not sure: if we have `1.1` and `1.2`. I believe what @krauselukas thought when coming up with the versioning schema is to use the "biggest" of the "minor" version. I'm not sure if `least` is the word that translates to that then...

Author: [Pessimistic Version Constraint](https://guides.rubygems.org/patterns/#pessimistic-version-constraint) is what I was talking about

Reviewer: We are making up a lot of stuff on the fly. I would not mention this to customers yet...

Author: Then I will mention that we don't enforce any specific version on the workflow configuration, that we'll default to the lowest minor version and that's it

---
layout: post

> title: Versioning in the SCM/CI Integration

Reviewer 2: ```suggestion
title: Versioning arrives to SCM/CI Integration
```
Just another suggestion.

> title: Versioning in the SCM/CI Integration

Author: I went for a variation on @hellcp-work's suggestion. WDYT?

> title: Versioning in the SCM/CI Integration

Reviewer: Sounds to me as if you are enabling larger changes in the SCM/CI workflow *right now*

category: development
---

Hello everyone!

The goal of this post is to explain how we're going to release new features to the SCM/CI integration.

# Versioning the Workflow Configuration

We are introducing the versioning as a way to release new features in the SCM/CI without breaking existing user's configuration. Those features are going to be released under a MAJOR.MINOR version scheme. You know the drill already: current version is 1.0, so the next minor version will be 1.1 and the next major version will probably be 2.0.

A major version update will introduce big features and may bring breaking changes, so it will require adapting your workflow configuration.

> A minor version update will only include small, backwards-compatible changes. If you don't touch anything, your workflow will still work as before. Only when you make some minor adjustments to your workflow configuration you'll benefit of a new feature.

Author: Well, it's what we wrote [on the documentation](https://github.com/openSUSE/obs-docu/pull/292). I thought we already agreed on that.

> A minor version update will only include small, backwards-compatible changes. If you don't touch anything, your workflow will still work as before. Only when you make some minor adjustments to your workflow configuration you'll benefit of a new feature.

Reviewer: Guess I missed that okay...

The workflow version can be specified as the top-level element of the workflow configuration yaml file. Like this:

    version: 1.0
    workflow:
        step:
            - link_package:
                source_project: GNOME:Factory
                source_package: gnome-shell
                target_project: home:jane:playground

> Right now, we don't enforce a specific version in the yaml configuration. So if you don't set any version, your Workflow configuraiton will default to the latest minor version (1.1 as of the release of this blog post). That way you can test those minor features at your own pace.

Reviewer: This sounds as if we are going to change this in the future. We wont...
